invisible_cities/cities/detsim_simulate_electrons_test.py

```diff
@@ -56,6 +56,13 @@ def test_generate_ionization_electrons(MChits_and_detsim_electron_simulation_par
     assert len(secondary_electrons) == len(energies)
     assert np.all(secondary_electrons >= 0)
 
+    isinteger = lambda value: int(value) == value
```

Comment:
Anything wrong with `float.is_integer`?

Reply:
This works if the values were `float`. `generate_ionization_electrons` and `drift_electrons` return `numpy.ndarray` of type `numpy.int64` values.

Reply:
Apologies if I'm being thick, but if `generate_ionization_electrons` returns an array of `int64`s, then I'm struggling to see the point of checking whether they are all integers. (Maybe I've been spending too much time with static type systems lately?)

(You mention `drift_electrons` in response to a comment about `isinteger` but I don't see how the two are related.)

Also, whilst we're here, the fact that line 63 is character-for-character identical to line 54 is a bit confusing. I appreciate that the energies are different, but using the same name in the same scope for a different set of energies promotes confusion.

Line 62 onwards seems to be completely independent of what precedes it (besides using the same fixture), which suggests that it belongs in a separate test.

Reply:
For the integers I was following the comments: https://github.com/nextic/IC/pull/695#discussion_r467050629, https://github.com/nextic/IC/pull/695#discussion_r467050093. Maybe I have not understand it properly, @gonzaponte.

I do not know if it worths to split the tests for null `energies` in `generate_ionization_electrons`; for null and infinite `lifetime` in `drift_electrons` and for null diffusions and negative `zs` in `diffuse_electrons`...

Reply:
I don't understand the comments you link: If the type is an integer type, then the values are guaranteed to be integers! If the type is `float` then you should use `float.is_integer` rather than baking your own. The only situation in which it might make sense is if *sometimes* the values are `float`s and *sometimes* `int`s ... but you claim that they are *always* `int64`, so I'm lost.

As for splitting the tests: tests should be as orthogonal, independent and short as is *reasonably* possible (I'll be happy to explain this at great length, if you wish, but this is not the place for it). As ever, don't take this as dogma, but as a guiding principle. These two bits of the test are idependent. Is there any reason why they should be grouped under the same test? I don't see the reason, but I'd be happy for it to be pointed out to me.

Reply:
Sorry, I'm still catching up with this. I haven't been able to keep an eye on this yesterday.

> Maybe I have not understand it properly, @gonzaponte.

No, I think you understood correctly. I meant that we want to check the value, not the type.

> If the type is an integer type, then the values are guaranteed to be integers! 

Indeed. But the function returning an int-array type is just an implementation choice (and more intuitive for the human eye). Everything should work the same whether we have an int-type array or a float-type array holding integers only (I think). I suggested value-based tests following the duck typing paradigm. I want to make sure we don't return non-integer values, not non-integer types. Does it sound right?

As for the splitting the tests, yes, please. Make an individual test for each feature/behavior as much as possible.

Reply:
So the value of the `isinteger` check is protection against the return type of `generate_ionization_electrons` changing to some kind of `float` which would then admit non-integer values, which might the actually appear in the array?

Hmm, OK ...

> I suggested value-based tests following the duck typing paradigm.

I wonder whether that's taking it too far. Just because Duck Typing works well in Python in many cases, and we should use it to our advantage when we can, that doesn't mean that we should religiously try to force it in everywhere: in some cases Duck Typing is simply not the best option. If it is nonsensical for these values to be non-integers, I think that using the (non-Duck) type system to say so, is probably more appropriate.

How you express this idea with sufficient generality in Python's contrete type system is another question. There's `numpy.integer` but unfortunately the builtin `int` is not related to it.

[As a fun aside, let's nod to Purescript's type system, which would allow us to specify concrete type constraints that nail down the [exact algebraic properties](https://a-guide-to-the-purescript-numeric-hierarchy.readthedocs.io/en/latest/introduction.html) that these "numbers" should have! In other words, it's a static-formal-duck-type-system[*] which allows you to say exactly what you want with remarkable precision.

[*] By formal-duck I mean that it not only has to quack, but the quack actually has to have the intended meaning.]

Reply:
Ok, so you suggest we stick to return integer values and check for the type in the tests?

Reply:
Don't take my word as gospel, especially in this case where I've only scratched the surface (so, perhaps if I look at more context, my opinion will change), but a type check looks reasonable to me at the moment.

The problem is generality. How about something like `is{instance/subclass}(..., (np.integer, int))`?

Reply:
From your previous explanation, it sounds reasonable. We are definitely working with something that doesn't make sense in non-integer values (you cannot have 3.5 electrons), so it makes sense to restrict the type. Now, within the type `integer` we don't care which one it is as long as it does not limit the values (we don't want `np.int8` or `np.int16`). That aside, should we do something like `"int" in array.dtype.name`?

Reply:
> `"int" in array.dtype.name`

Bleargh!

It's most unfortunate that the use of strings to represent more meaningful and structured information, has gained such popularity in Python culture: it's quite an abominable practice. I see the attraction here, as it lets you write something that isn't obviously incorrect at first blush in not too many characters, but ... we don't care about a relationship between sequences of characters, we care about types! And types are first-class objects in Python!

Can I encourage you to eschew text-pattern-matching and tend towards using objects, and their meaningful relationships and/or identities?

How about `issubclass(arr.dtype.type, (np.integer, int))`?

Reply:
> Bleargh!

I know, I know.. But I thought there was no way to include `uint`s in the equation otherwise. Fortunately 

> How about issubclass(arr.dtype.type, (np.integer, int))?

works, so let's go for it.

Reply:
> Indeed. But the function returning an int-array type is just an implementation choice (more intuitive for the human mind). Everything should work the same whether we have an int-type array or a float-type array holding integers only (I think). I suggested value-based tests following the duck typing paradigm. I want to make sure we don't return non-integer values, not non-integer types. Does it sound right?

No, in any case they should be int-types since in `drift_electrons` they enter as a `size` parameter and in `diffuse_electrons` in `np.repeat` as the repeats. So a float-type will not work even if it is an integer-value.

Reply:
Correct. I had tried with a float value in `np.repeat` and it worked, but not with a `np.ndarray` containing floats.